File: this_is_coding_test/7-binary_search/slicing_cake.cpp
#include <stdio.h>

using namespace std;

int N, M, minVal = 1000000000, maxHeight;
int cakeLen[1000000];

int sliceCake(int height) {
    int sum = 0;
    
    for (int i = 0; i < N; ++i) {
        int slicedLen = cakeLen[i] - height;

        if (slicedLen > 0) {
            sum += slicedLen;
        } 
    }

    return sum;
}

void findHeight(int start, int end) {
    int mid = (start + end) / 2, slicedLen;

    if (start > end) return;

    slicedLen = sliceCake(mid);

    if (slicedLen < M) {
        findHeight(start, mid - 1);
    } else {
        minVal = mid;

        findHeight(mid + 1, end);
    }

}

int main() {
    int tmp;

    scanf("%d %d", &N, &M);

    for (int i = 0; i < N; ++i) {
        scanf("%d", &tmp);
        
        cakeLen[i] = tmp;

        if (tmp > maxHeight) {
            maxHeight = tmp;
        }
    }

    findHeight(0, maxHeight - 1);

    printf("%d", minVal);

    return 0;
}

File: this_is_coding_test/7-binary_search/slicing_cake.py
from symbol import sliceop


N, M = map(int, input().split())
cake_list = list(map(int, input().split()))

max_in_list = max(cake_list)
min_val = 0


def slice_cake(h):
    sliced_height = 0

    for height in cake_list:
        if (height > h):
            sliced_height += height - h

    return sliced_height


def binary_search(start, end):
    if (start > end):
        return None

    global min_val

    mid = (start + end) // 2

    sliced_height = slice_cake(mid)

    if (sliced_height == M):
        min_val = mid
    elif (sliced_height > M):
        min_val = mid
        binary_search(mid + 1, end)
    else:
        binary_search(start, mid - 1)


binary_search(0, max_in_list)
print(min_val)
